listenbrainz/webserver/views/playlist_api.py

```diff
@@ -0,0 +1,502 @@
+PLAYLIST_RELEASE_URI_PREFIX = "https://musicbrainz.org/release/"
+PLAYLIST_URI_PREFIX = "https://listenbrainz.org/playlist/"
+PLAYLIST_EXTENSION_URI = "https://musicbrainz.org/doc/jspf#playlist"
+PLAYLIST_TRACK_EXTENSION_URI = "https://musicbrainz.org/doc/jspf#track"
+RECORDING_LOOKUP_SERVER_URL = "https://labs.api.listenbrainz.org/recording-mbid-lookup/json"
+MAX_RECORDINGS_PER_ADD = 100
+
+
+def _parse_boolean_arg(name, default=None):
+    value = request.args.get(name)
+    if not value:
+        return default
+
+    value = value.lower()
+    if value not in ["true", "false"]:
+        raise APIBadRequest("Invalid %s argument: %s. Must be 'true' or 'false'" % (name, value))
+
+    return True if value == "true" else False
+
+
+def validate_playlist(jspf, require_title=True):
+    """
+        Given a JSPF dict, ensure that title is present and that if tracks are present
+        they have valid URIs or MBIDs specified. If any errors are found 400 is raised.
+    """
+
+    if require_title:
+        try:
+            title = jspf["playlist"]["title"]
+            if not title:
+                raise KeyError
+        except KeyError:
+            log_raise_400("JSPF playlist must contain a title element with the title of the playlist.")
+
+    if 'track' not in jspf:
+        return
+
+    for i, track in enumerate(jspf['track']):
+        try:
+            recording_uri = track["identifier"]
+            if not recording_uri:
+                raise KeyError
+        except KeyError:
+            log_raise_400("JSPF playlist track %d must contain an identifier element with recording MBID." % i)
+
+        if recording_uri.startswith(PLAYLIST_TRACK_URI_PREFIX):
+            recording_mbid = recording_uri[len(PLAYLIST_TRACK_URI_PREFIX):]
+        else:
+            log_raise_400("JSPF playlist track %d identifier must have the namespace '%s' prepended to it." %
+                          (i, PLAYLIST_TRACK_URI_PREFIX))
+
+        if not is_valid_uuid(recording_mbid):
+            log_raise_400("JSPF playlist track %d does not contain a valid track identifier field." % i)
+
+
+def serialize_jspf(playlist: Playlist):
+    """
+        Given a playlist, return a properly formated dict that can be passed to jsonify.
+        TODO: Add tests for newly added fields.
+              Add collaborators
+    """
+
+    pl = {"creator": playlist.creator,
+          "title": playlist.name,
+          "identifier": PLAYLIST_URI_PREFIX + str(playlist.mbid),
+          "date": playlist.created.replace(tzinfo=datetime.timezone.utc).isoformat(),
+    }
+    if playlist.description:
+        pl["annotation"] = playlist.description
+
+    extension = {"public": "true" if playlist.public else "false"}
+    extension["creator_id"] = playlist.creator_id
+    if playlist.created_for_id:
+        extension['created_for'] = playlist.created_for
+    if playlist.copied_from_id:
+        extension['copied_from'] = PLAYLIST_URI_PREFIX + str(playlist.copied_from_id)
```

Comment:
This currently return the DB id rather than the MBID of the playlist.

I also wonder if created_for currently returns a name string rather than a user id.

Reply:
when we load a playlist we should populate a field `copied_from` which is the mbid of the playlist referred to by `copied_from_id`